src/parser.js

```diff
@@ -0,0 +1,197 @@
+const { ...structs }			= require('./structs.js');
+
+const delay				= (ms) => new Promise(f => setTimeout(f, ms));
+
+const STATE_GET_HEADER			= 0;
+const STATE_GET_PAYLOAD			= 1;
+
+
+class ParserError extends LairClientError {
+    [Symbol.toStringTag]		= ParserError.name;
+}
+
+
+class MessageParser extends stream.Duplex {
+    constructor () {
+	super()
+
+	this.msg_count			= 0;
+
+	this.state			= STATE_GET_HEADER;
+
+	this.need_more_buffer		= true;
+	this._bufferedBytes		= 0;
+	this._buffers			= [];
+	this._payloadLength		= 0;
+	this.current_msg		= null;
+	this.next_payload		= null;
+    }
+
+    _write ( buf, encoding, cb ) {
+	if ( ! (buf instanceof Uint8Array) )
+	    throw new ParserError(`Cannot consume type '${typeof buf}'`);
+
+	log.silly("Incoming chunk (%s): %s", buf.length, buf.toString('hex') );
+	this._bufferedBytes	       += buf.length;
+	this._buffers.push( buf );
+	this.need_more_buffer		= false;
+
+	this._processBuffers().catch(console.error);
+
+	return cb() || true;
+    }
+
+    async _processBuffers () {
+	await delay(0);
+
+	do {
+	    log.silly("State maching loop: %s (%s bytes avaiable)", this.state, this._bufferedBytes );
+	    switch ( this.state ) {
+	    case STATE_GET_HEADER:
+		this._parseHeader();
+		break;
+	    case STATE_GET_PAYLOAD:
+		this._parsePayload();
+		break;
+	    default:
+		this.need_more_buffer	= true;
+		break;
+	    }
+	    log.silly("End of loop state: %s (more? %s)", this.state, this.need_more_buffer );
+	} while (this.need_more_buffer === false);
+    }
+
+    _consume ( n ) {
+	this._bufferedBytes	       -= n;
+
+	if ( n === this._buffers[0].length )
+	    return this._buffers.shift();
+
+	if ( n < this._buffers[0].length ) {
+	    const buf			= this._buffers[0];
+	    this._buffers[0]		= buf.slice( n );
+	    return buf.slice( 0, n );
+	}
+
+	const dst			= Buffer.allocUnsafe( n );
+
+	do {
+	    const buf			= this._buffers[0];
+	    const offset		= dst.length - n;
+
+	    if ( n >= buf.length ) {
+		dst.set( this._buffers.shift(), offset );
+	    } else {
+		dst.set( new Uint8Array(buf.buffer, buf.byteOffset, n), offset );
+		this._buffers[0]	= buf.slice( n );
+	    }
+
+	    n -= buf.length;
+	} while (n > 0);
+
+	return dst;
+    }
+
+    _parseHeader () {
+	if ( this._bufferedBytes < HEADER_SIZE ) {
+	    this.need_more_buffer	= true;
+	    return;
+	}
+	log.info("Parsing (%s) header from %s available bytes", HEADER_SIZE, this._bufferedBytes );
+
+	let header_bytes		= this._consume( HEADER_SIZE );
+
+	let payload_promise		= new Promise((f,r) => {
+	    this.next_payload		= payload => {
+		this.next_payload	= null;
+		f(payload);
+	    }
+	});
+
+	const wt_id			= header_bytes.readUInt32LE(4);
+	const wt_cls			= structs[wt_id];
+	this.current_msg		= {
+	    "length":		header_bytes.readUInt32LE() - HEADER_SIZE,
+	    "wire_type":	wt_cls.name,
+	    "wire_type_id":	wt_id,
+	    "wire_type_class":	wt_cls,
+	    "id":		parseInt( header_bytes.readBigUInt64LE(8) ),
```

Comment:
I know you are generating 32 bit integers for request message ids, but the remote may not for say the the unlock passphrase request - might this parseInt fail if the number is > 2**32?

Reply:
Yes, it would fail.  My plan is to change the handling of message IDs so that they remain in byte form.  Not sure if that will happen before the release, but I'll make an issue for it so we don't forget.

Reply:
Actually, I am thinking of the wrong limitation here.  I had this problem with 8 byte lengths because most Javascript functions do not handle `BigInt`.  However, the message ID can remain a `BigInt` the whole time so I should be able to fix this before the release.

Reply:
Fixed in https://github.com/Holo-Host/lair-client-js/pull/1/commits/b66a0a81a7b32bcc5ab2e98c933bd0eab2c97848